internal/lifecycle/readme.go

```diff
@@ -0,0 +1,205 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lifecycle
+
+import (
+	"bufio"
+	"fmt"
+	"github.com/GoogleCloudPlatform/serverless-sample-tester/internal/util"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+)
+
+const (
+	// The tag that should appear immediately before code blocks in a README to indicate that the enclosed commands
+	// are to be used by this program for building and deploying the sample.
+	codeTag = "{sst-run-unix}"
+
+	// A non-quoted backslash in bash at the end of a line indicates a line continuation from the current line to the
+	// next line.
+	bashLineContChar = '\\'
+)
+
+var (
+	gcloudCommandRegexp   = regexp.MustCompile(`^gcloud\b`)
+	cloudRunCommandRegexp = regexp.MustCompile(`\brun\b`)
+
+	gcrURLRegexp = regexp.MustCompile(`gcr.io/.+/\S+`)
+
+	mdCodeFenceStartRegexp = regexp.MustCompile("^\\w*`{3,}[^`]*$")
+
+	errNoREADMECommandsFound = fmt.Errorf("[lifecycle.parseREADME]: no commands found")
+)
+
+// parseREADME parses a README file with the given name. It reads terminal commands surrounded by one of the codeTags
+// listed above and loads them into a Lifecycle. In the process, it replaces the Cloud Run service name and Container
+// Registry tag with the provided inputs.
+func parseREADME(filename, serviceName, gcrURL string) (Lifecycle, error) {
+	file, err := os.Open(filename)
+	if err != nil {
+		return nil, fmt.Errorf("[lifecycle.parseREADME] os.Open %s: %w", filename, err)
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	codeBlocks, err := codeBlocks(scanner)
+	if err != nil {
+		return nil, fmt.Errorf("[lifecycle.parseREADME] extracting code blocks out of %s: %w", filename, err)
+	}
+
+	lifecycle, err := codeBlocksTolifecycle(codeBlocks, serviceName, gcrURL)
+	if err != nil {
+		return lifecycle, fmt.Errorf("[lifecycle.parseREADME] transforming code blocks in %s to executable "+
+			"commands: %w", filename, err)
+	}
+
+	if len(lifecycle) == 0 {
+		return lifecycle, errNoREADMECommandsFound
+	}
+
+	return lifecycle, nil
+}
+
+// codeBlocks extracts code blocks out of a bufio.Scanner that's reading from a Markdown file immediately prefaced with
+// a line containing codeTag. It returns an 2d slice of code blocks, each containing an array of lines contained within
+// that code block.
+func codeBlocks(scanner *bufio.Scanner) ([][]string, error) {
+	var blocks [][]string
+
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if strings.Contains(line, codeTag) {
+			if s := scanner.Scan(); !s {
+				if err := scanner.Err(); err != nil && !s {
```

Comment:
nit: I think `!s` is already true if we're in this section?
```suggestion
				if err := scanner.Err(); err != nil {
```

Reply:
Resolved, thank you